File: tools/replace4Macro_matlab_python/replace4Macro.m
% This script is designed to replicate the Replace4Macro that is used in
% BodyBuilder. The original code can be found from our website
% www.vicon.com/downloads under the heading "Models and Scripts", in the
% download called "BodyBuilder Example Models".
%
% Written by Nev Pires, 12/2/2020
%
% This code will work if one marker on a rigid body segment is missing for
% a few frames in the trial. If two or more markers on the rigid body
% segment are missing in the same frame/s, they will not be replaced. If a
% marker is missing for the entire trial, please use the companion code
% "replaceMissingMarker.m"
%
% The inputs are the Subject Name: as a string which is the same as the VSK
% name;
% p1, p2, p3 and p4 are the markers on the rigid body segment.
% These should be entered as strings, and they need to match the marker names in the VSK.
% If one out of the four markers are missing (but is present in at least one frame), 
% this code will replace the
% missing marker.

%% Clear the command window
clear; clc;

%% Connect to Nexus
vicon = ViconNexus();

%% Create the dailog box for the subject name, and the four markers on the rigid body cluster
inputs = inputdlg({'Subject Name', 'Marker1', 'Marker2', 'Marker3', 'Marker4'}, 'Rigid Body Marker Clusters');

%% Parse the inputs into their respective variables 
subject = inputs{1};

marker_names = sort(inputs(2:end));

%% See if the subject name is in the list of subjects from Nexus

subject_list = vicon.GetSubjectNames();

if ~any(strcmp(subject, subject_list))
    questdlg('The Subject Name does not match the VSK. Please enter the correct Subject Name', ...
        '','Ok','Ok');
end

%% Get the Marker Information

full_marker_list = vicon.GetMarkerNames(subject);

found = logical(true(1,length(marker_names)));
for marker = 1:length(marker_names)
    found(marker) = false;
    for marker_list_name = 1:length(full_marker_list)
        if strcmp(full_marker_list(marker_list_name), marker_names{marker}) == 1
            found(marker) = true;
        end
    end
end

%% Check if Trial has all four markers present

trajExists = logical(true(1,length(marker_names)));

for marker = 1:length(marker_names)
    if found(marker) == true
        trajExists(marker) = vicon.HasTrajectory(subject, marker_names{marker});
    else
        error('A marker does not exist in the VSK')
    end
end

%% Get Marker Data

for marker = 1:length(marker_names)

    [x.(marker_names{marker}), y.(marker_names{marker}), z.(marker_names{marker}), e.(marker_names{marker})] = vicon.GetTrajectory(subject, marker_names{marker});
     
    for frame = 1:length(x.(marker_names{marker}))
        if e.(marker_names{marker})(frame) == false
            x.(marker_names{marker})(frame) = NaN;
            y.(marker_names{marker})(frame) = NaN;
            z.(marker_names{marker})(frame) = NaN;
        end
    end

    marker_data.(marker_names{marker}) = [x.(marker_names{marker}); y.(marker_names{marker}); z.(marker_names{marker})];

end

%% Create Nexus trajectory objects for each marker and populate them with the marker data that has already been imported

marker1 = NexusTrajectory(subject);
marker1.SetPosition(marker_data.(marker_names{1}));

marker2 = NexusTrajectory(subject);
marker2.SetPosition(marker_data.(marker_names{2}));

marker3 = NexusTrajectory(subject);
marker3.SetPosition(marker_data.(marker_names{3}));

marker4 = NexusTrajectory(subject);
marker4.SetPosition(marker_data.(marker_names{4}));

%% Define Segments if a single marker on the 4 marker cluster is missing

s234 = NexusSegment(subject); % Creates Nexus segment object in Matlab
s234.Populate(marker3, marker2 - marker3, marker3 - marker4, 'xyz'); % Creates a segment with markers 2, 3, 4 as marker 1 is missing

s134 = NexusSegment(subject); % Creates Nexus segment object in Matlab
s134.Populate(marker3, marker1 - marker3, marker3 - marker4, 'xyz'); % Creates a segment with markers 1, 3, 4 as marker 2 is missing

s124 = NexusSegment(subject); % Creates Nexus segment object in Matlab
s124.Populate(marker2, marker1 - marker2, marker2 - marker4, 'xyz'); % Creates a segment with markers 1, 2, 4 as marker 3 is missing

s123 = NexusSegment(subject); % Creates Nexus segment object in Matlab
s123.Populate(marker2, marker1 - marker2, marker2 - marker3, 'xyz'); % Creates a segment with markers 1, 2, 3 as marker 4 is missing

%% Localise the missing marker in the corresponding coordinate system (created from the 3 remaining markers)

p1_local = s234.LocalisePoint(marker1); % Localise p1 in segment s234
p2_local = s134.LocalisePoint(marker2); % Localise p2 in segment s134
p3_local = s124.LocalisePoint(marker3); % Localise p3 in segment s124
p4_local = s123.LocalisePoint(marker4); % Localise p4 in segment s123

%% Calculate the mean of the localised markers

p1mean = mean(p1_local','omitnan')'; % Calculates the average of p1 local point
p2mean = mean(p2_local','omitnan')'; % Calculates the average of p2 local point
p3mean = mean(p3_local','omitnan')'; % Calculates the average of p3 local point
p4mean = mean(p4_local','omitnan')'; % Calculates the average of p4 local point

%% Create Nexus Trajectory objects for each marker

p1 = NexusTrajectory(subject);
p2 = NexusTrajectory(subject);
p3 = NexusTrajectory(subject);
p4 = NexusTrajectory(subject);

%% Globalize each point

p1.SetPosition(s234.GlobalisePoint(p1mean));
p2.SetPosition(s134.GlobalisePoint(p2mean));
p3.SetPosition(s124.GlobalisePoint(p3mean));
p4.SetPosition(s123.GlobalisePoint(p4mean));

%% Extract the position for each marker from the Nexus object

position_outputs.(marker_names{1}) = p1.Position;
position_outputs.(marker_names{2}) = p2.Position;
position_outputs.(marker_names{3}) = p3.Position;
position_outputs.(marker_names{4}) = p4.Position;

%% Replace missing frames (NaN's) in original data with the calculated data

for marker = 1:length(marker_names)
    for frame = 1:length(marker_data.(marker_names{marker}))
        if all(isnan(marker_data.(marker_names{marker})(:,frame)))
            marker_data.(marker_names{marker})(:,frame) = position_outputs.(marker_names{marker})(:,frame);
        else
            marker_data.(marker_names{marker})(:,frame) = marker_data.(marker_names{marker})(:,frame);
        end
    end
end

%% Determine if marker exists for the frame

for marker = 1:length(marker_names)
    for frame = 1:length(marker_data.(marker_names{marker}))
        if all(isnan(marker_data.(marker_names{marker})(:,frame)))
            exists_new.(marker_names{marker})(frame) = false;
        else
            exists_new.(marker_names{marker})(frame) = true;
        end
    end
end

%% Set Trajectories

for marker = 1:length(marker_names)

    vicon.SetTrajectory(subject, marker_names{marker},...
        marker_data.(marker_names{marker})(1,:),...
        marker_data.(marker_names{marker})(2,:),...
        marker_data.(marker_names{marker})(3,:),...
        exists_new.(marker_names{marker}))

end

File: tools/replace4Macro_matlab_python/replace4Macro.py
# This script is designed to replicate the Replace4Macro that is used in
# BodyBuilder. The original code can be found from our website
# www.vicon.com/downloads under the heading "Models and Scripts", in the
# download called "BodyBuilder Example Models".
#
# Written by Nev Pires, 12/2/2020
#
# This code will work if one marker on a rigid body segment is missing for
# a few frames in the trial. If two or more markers on the rigid body
# segment are missing in the same frame/s, they will not be replaced. If a
# marker is missing for the entire trial, please use the companion code
# "replaceMissingMarker.py"
#
# The inputs are the p1, p2, p3 and p4 are the markers on the rigid body segment.
# These should be entered as strings.
# The subject name must be selected from the drop down menu.
# If one out of the four markers are missing (but is present in at least one frame),
# this code will replace the missing marker.
#
# This script requires the following modules:
#
# - numpy
# - tkinter
# - viconnexusapi
# - viconnexusutils

from viconnexusapi import ViconNexus
import numpy as np
from viconnexusutils import NexusSegment
from viconnexusutils import NexusTrajectory
import tkinter as tk
from tkinter import messagebox
import sys

# Connects to Nexus
vicon = ViconNexus.ViconNexus()

# Gets the list of subject names
subject_list = vicon.GetSubjectNames()

# Get Start and End frames for the entire trial
start_frame, end_frame = vicon.GetTrialRegionOfInterest()


class Input(tk.Frame):

    # Creates the input box where the 4 markers on the rigid body segment are to be entered and the subject name is to be selected from the dropdown menu

    def __init__(self, parent):
        tk.Frame.__init__(self, parent)
        self.parent = parent

        # Creates the dropdown menu with the subject names listed

        self.subject_selection = tk.StringVar()
        self.subject_selection.set(subject_list[0])

        # Creates 4 input boxes that correspond with the 4 markers that are on the rigid body segment

        self.marker_1_label = tk.Label(root, text="Marker 1: ")
        self.marker_1_entry = tk.Entry(root)

        self.marker_2_label = tk.Label(root, text="Marker 2: ")
        self.marker_2_entry = tk.Entry(root)

        self.marker_3_label = tk.Label(root, text="Marker 3: ")
        self.marker_3_entry = tk.Entry(root)

        self.marker_4_label = tk.Label(root, text="Marker 4: ")
        self.marker_4_entry = tk.Entry(root)

        # Creates the dropdown menu with the subject names listed. A default subject is chosen if multiple subjects are in the trial

        self.ideal_label = tk.Label(root, text="Subject Name: ")
        self.ideal_entry = tk.OptionMenu(root, self.subject_selection, *subject_list)

        # Creates a button to close the window and continue

        self.submit_button = tk.Button(text="Submit", command=self.close_window)

        # Creates the location of the text and input boxes for the dialog box

        self.marker_1_label.grid(row=0, column=0)
        self.marker_1_entry.grid(row=0, column=1)

        self.marker_2_label.grid(row=1, column=0)
        self.marker_2_entry.grid(row=1, column=1)

        self.marker_3_label.grid(row=2, column=0)
        self.marker_3_entry.grid(row=2, column=1)

        self.marker_4_label.grid(row=3, column=0)
        self.marker_4_entry.grid(row=3, column=1)

        self.ideal_label.grid(row=4, column=0)
        self.submit_button.grid(columnspan=2, row=5, column=0)

        self.ideal_entry.grid(row=4, column=1)

    # function to close the window

    def close_window(self):
        # global name
        # global ideal_type
        self.marker1 = self.marker_1_entry.get()
        self.marker2 = self.marker_2_entry.get()
        self.marker3 = self.marker_3_entry.get()
        self.marker4 = self.marker_4_entry.get()
        self.subject = self.subject_selection.get()
        # self.destroy()
        self.quit()


# Executes the above function and class

if __name__ == '__main__':
    root = tk.Tk()
    app = Input(root)
    root.mainloop()
    # Note the returned variables here
    # They must be assigned to external variables
    # for continued use
    marker1 = app.marker1
    marker2 = app.marker2
    marker3 = app.marker3
    marker4 = app.marker4
    subject = app.subject
    # Should only need root.destroy() to close down tkinter
    # But need to handle user cancelling the form instead
    try:
        root.destroy()
    except:
        sys.exit(1)

# Check to see if markers exist in the subject list

marker_list = sorted([marker1, marker2, marker3, marker4])
full_marker_list = vicon.GetMarkerNames(subject)

found = [marker in full_marker_list for marker in marker_list]
trajExists = [[] for item in range(len(found))]

for marker in range(len(marker_list)):
    if found[marker]:
        trajExists[marker] = vicon.HasTrajectory(subject, marker_list[marker])
    else:
        messagebox.showinfo('ERROR', 'Error: Marker ' + ' ' + marker_list[marker] + ' ' + ' does not exist in VSK')
        sys.exit('Error: Marker ' + ' ' + marker_list[marker] + ' ' + ' does not exist in VSK')

markers = {names: NexusTrajectory.NexusTrajectory(subject) for names in marker_list}
_ = {names: markers[names].Read(names, vicon) for names in marker_list}
markers = {names: np.array(markers[names].Position()) for names in marker_list}

def CreateSegment(subject, origin, p1, p2, p3):
    """Creates and populates the Nexus Segment. The segment has a rotation of xyz"""
    segment = NexusSegment.NexusSegment(subject)
    segment.Populate(origin, p1 - p2, p2 - p3, "xyz")

    return segment

# Create a dictionary for the permutations of the missing markers
segments = {"s123": {"segment": CreateSegment(subject, markers[marker_list[2]],
                                              markers[marker_list[1]], markers[marker_list[2]],
                                              markers[marker_list[3]]), marker_list[0]: markers[marker_list[0]]},
            "s023": {"segment": CreateSegment(subject, markers[marker_list[2]],
                                              markers[marker_list[0]], markers[marker_list[2]],
                                              markers[marker_list[3]]), marker_list[1]: markers[marker_list[1]]},
            "s013": {"segment": CreateSegment(subject, markers[marker_list[1]],
                                              markers[marker_list[0]], markers[marker_list[1]],
                                              markers[marker_list[3]]), marker_list[2]: markers[marker_list[2]]},
            "s012": {"segment": CreateSegment(subject, markers[marker_list[1]],
                                              markers[marker_list[0]], markers[marker_list[1]],
                                              markers[marker_list[2]]), marker_list[3]: markers[marker_list[3]]}}

# Create a dictionary for the segments based on the remaining markers
local_point = {name: np.array(segments[segment_name]["segment"].LocalisePoint(segments[segment_name][name]))
               for segment_name in segments.keys() for name in segments[segment_name].keys() if name != "segment"}

# Calculate the local point average
local_point_ave = {name: np.nanmean(local_point[name], axis=0) for name in local_point.keys()}

# Create Nexus Trajectory objects for each marker
p_markers = {name: NexusTrajectory.NexusTrajectory(subject) for name in local_point}

# Globalize each point
_ = {name: p_markers[name].SetPosition(segments[segment_name]["segment"].GlobalisePoint(local_point_ave[name]))
     for segment_name in segments.keys() for name in segments[segment_name].keys() if name != "segment"}

# Extract the position for each marker from the Nexus object
p_positions = {name: np.array(p_markers[name].Position()) for name in p_markers}

# Replace missing frames (NaN's) in original data with the calculated data
marker_data = {name: np.array([p_positions[name][frame] if all(np.isnan(markers[name][frame]))
               else markers[name][frame] for frame in range(start_frame-1, end_frame)]) for name in p_positions.keys()}

# Determine if marker exists or not for a specific frame
exists_new = {names: [False if all(np.isnan(frames)) else True for frames in marker_data[names]]
              for names in marker_list}

# setPosition = {names: markers[names].SetPosition(markerData[marker_list[names]]) for names in marker_list}
_ = {name: vicon.SetTrajectory(subject, name,
                               list(marker_data[name][:, 0]),
                               list(marker_data[name][:, 1]),
                               list(marker_data[name][:, 2]),
                               exists_new[name]) for name in marker_list}
